wasm/src/lsp/types.rs

```diff
@@ -0,0 +1,300 @@
+use std::cmp::PartialOrd;
+use std::path::PathBuf;
+use std::str::FromStr;
+use std::rc::Rc;
+
+use lsp_server::{ExtractError, Request, RequestId};
+
+use lsp_types::{
+    Position,
+    Range, SemanticTokenModifier, SemanticTokenType
+};
+
+use percent_encoding::percent_decode;
+use url::{Host, Url};
+
+use clvm_tools_rs::compiler::sexp::decode_string;
+use clvm_tools_rs::compiler::srcloc::Srcloc;
+
+lazy_static! {
+    pub static ref TOKEN_TYPES: Vec<SemanticTokenType> = {
+        vec![
+            SemanticTokenType::PARAMETER,
+            SemanticTokenType::VARIABLE,
+            SemanticTokenType::FUNCTION,
+            SemanticTokenType::MACRO,
+            SemanticTokenType::KEYWORD,
+            SemanticTokenType::COMMENT,
+            SemanticTokenType::STRING,
+            SemanticTokenType::NUMBER,
+        ]
+    };
+    pub static ref TOKEN_MODIFIERS: Vec<SemanticTokenModifier> = {
+        vec![
+            SemanticTokenModifier::DEFINITION,
+            SemanticTokenModifier::READONLY,
+            SemanticTokenModifier::DOCUMENTATION,
+        ]
+    };
+}
+
+pub const TK_PARAMETER_IDX: u32 = 0;
+pub const TK_VARIABLE_IDX: u32 = 1;
+pub const TK_FUNCTION_IDX: u32 = 2;
+pub const TK_MACRO_IDX: u32 = 3;
+pub const TK_KEYWORD_IDX: u32 = 4;
+pub const TK_COMMENT_IDX: u32 = 5;
+pub const TK_STRING_IDX: u32 = 6;
+pub const TK_NUMBER_IDX: u32 = 7;
+
+pub const TK_DEFINITION_BIT: u32 = 0;
+pub const TK_READONLY_BIT: u32 = 1;
+
+pub struct ToFilePathErr;
+
+// Note: to_file_path is only present on native builds, but we're building to
+// wasm.
+//
+// The code isn't ungeneric, so we'll express it here.
+pub trait HasFilePath {
+    fn our_to_file_path(&self) -> Result<PathBuf, ToFilePathErr>;
+}
+
+fn file_url_segments_to_pathbuf(
+    host: Option<Vec<u8>>,
+    segments: std::str::Split<'_, char>,
+) -> Result<PathBuf, ToFilePathErr> {
+    if host.is_some() {
+        return Err(ToFilePathErr);
+    }
+
+    let mut bytes = Vec::new();
+
+    for segment in segments {
+        bytes.push(b'/');
+        bytes.extend(percent_decode(segment.as_bytes()));
+    }
+
+    // A windows drive letter must end with a slash.
+    if bytes.len() > 2
+        && matches!(bytes[bytes.len() - 2], b'a'..=b'z' | b'A'..=b'Z')
+        && matches!(bytes[bytes.len() - 1], b':' | b'|')
+    {
+        bytes.push(b'/');
+    }
+
+    let path = PathBuf::from(decode_string(&bytes));
+
+    debug_assert!(
+        path.is_absolute(),
+        "to_file_path() failed to produce an absolute Path"
+    );
+
+    Ok(path)
+}
+
+impl HasFilePath for Url {
+    fn our_to_file_path(&self) -> Result<PathBuf, ToFilePathErr> {
+        if let Some(segments) = self.path_segments() {
+            let host = match self.host() {
+                None | Some(Host::Domain("localhost")) => None,
```

Comment:
Does this mean `localhost` URLs are illegal? Is 127.0.0.1 illegal too?

Reply:
It's kind of an incidental thing (and also someone else's code) but my interpretation is that they're giving a run at keeping localhost urls from being treated as local file paths.

This code is really incidental, and I chose it because it was self contained (and expedient to import).  It's 100% bad that we have a mismatch between Uri space and file space baked into the lsp protocol but I can't change that.